File: 006-react-portrait/src/services/comput3Api.js
import axios from 'axios';
import config from './config';

class Comput3API {
  /**
   * Client for interacting with Comput3 API
   * @param {string} apiKey - Comput3 API key
   */
  constructor(apiKey) {
    this.apiKey = apiKey;
    // Use relative path for the API URL to work with the proxy
    this.apiUrl = '/api/v0/workloads';
    
    if (!apiKey) {
      console.error('🔑 C3 API key is not provided. Please set your REACT_APP_C3_API_KEY in .env file.');
    }
  }
  
  /**
   * Get the headers for API requests
   * @returns {Object} Headers for API requests
   */
  getHeaders() {
    return {
      'accept': '/',
      'accept-language': 'en,en-US;q=0.9',
      'content-type': 'application/json',
      'origin': 'https://launch.comput3.ai',
      'referer': 'https://launch.comput3.ai/',
      'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36',
      'x-c3-api-key': this.apiKey
    };
  }
  
  /**
   * Get all running workloads from Comput3 API
   * @returns {Promise<Array>} List of running workloads
   */
  async getRunningWorkloads() {
    if (!this.apiKey) {
      console.error('❌ Cannot get workloads: C3 API key is missing');
      return [];
    }
    
    try {
      const response = await axios.post(
        this.apiUrl,
        { running: true },
        { headers: this.getHeaders() }
      );
      
      if (response.status !== 200) {
        console.error(`❌ Failed to get workloads: ${response.status} - ${response.data}`);
        return [];
      }
      
      const workloads = response.data;
      console.log(`🔍 Found ${workloads.length} running workloads`);
      return workloads;
    } catch (error) {
      console.error(`❌ Error getting workloads: ${error.message}`);
      return [];
    }
  }
  
  /**
   * Get a running media instance if available
   * @returns {Promise<Object|null>} Media instance or null if none found
   */
  async getMediaInstance() {
    const workloads = await this.getRunningWorkloads();
    
    // Filter for media instances
    const mediaInstances = workloads.filter(w => w.type && w.type.startsWith('media'));
    
    if (mediaInstances.length === 0) {
      console.warn('⚠️ No running media instances found');
      return null;
    }
    
    // Return the first media instance
    const instance = mediaInstances[0];
    console.log(`✅ Found media instance: ${instance.node} (type: ${instance.type})`);
    return instance;
  }
  
  /**
   * Get the ComfyUI URL for a running media instance
   * @returns {Promise<string|null>} ComfyUI URL or null if no instance found
   */
  async getComfyuiUrl() {
    const instance = await this.getMediaInstance();
    
    if (!instance || !instance.node) {
      return null;
    }
    
    const node = instance.node;
    const comfyuiUrl = `https://ui-${node}`;
    console.log(`🌐 ComfyUI URL: ${comfyuiUrl}`);
    return comfyuiUrl;
  }
}

export default Comput3API; 

File: 008-text-to-video-script/comput3_api.py
import requests
import json
import logging
from typing import Dict, Optional, List, Any

logger = logging.getLogger(__name__)

class Comput3API:
    """Client for interacting with Comput3 API"""
    
    def __init__(self, api_key: str):
        """Initialize with Comput3 API key"""
        self.api_key = api_key
        self.api_url = "https://api.comput3.ai/api/v0/workloads"
        
        if not api_key:
            logger.error("🔑 C3 API key is not provided. Please set your C3_API_KEY in .env file.")
        
    def get_headers(self) -> Dict[str, str]:
        """Get the headers for API requests"""
        return {
            "accept": "/",
            "accept-language": "en,en-US;q=0.9",
            "content-type": "application/json",
            "origin": "https://launch.comput3.ai",
            "referer": "https://launch.comput3.ai/",
            "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36",
            "x-c3-api-key": self.api_key
        }
    
    def get_running_workloads(self) -> List[Dict[str, Any]]:
        """Get all running workloads from Comput3 API"""
        if not self.api_key:
            logger.error("❌ Cannot get workloads: C3 API key is missing")
            return []
        
        try:
            response = requests.post(
                self.api_url,
                headers=self.get_headers(),
                json={"running": True}
            )
            
            if response.status_code != 200:
                logger.error(f"❌ Failed to get workloads: {response.status_code} - {response.text}")
                return []
            
            workloads = response.json()
            logger.info(f"🔍 Found {len(workloads)} running workloads")
            return workloads
            
        except Exception as e:
            logger.error(f"❌ Error getting workloads: {str(e)}")
            return []
    
    def get_media_instance(self) -> Optional[Dict[str, Any]]:
        """Get a running media instance if available"""
        workloads = self.get_running_workloads()
        
        # Filter for media instances
        media_instances = [w for w in workloads if w.get("type", "").startswith("media")]
        
        if not media_instances:
            logger.warning("⚠️ No running media instances found")
            return None
        
        # Return the first media instance
        instance = media_instances[0]
        logger.info(f"✅ Found media instance: {instance.get('node')} (type: {instance.get('type')})")
        return instance
    
    def get_comfyui_url(self) -> Optional[str]:
        """Get the ComfyUI URL for a running media instance"""
        instance = self.get_media_instance()
        
        if not instance or "node" not in instance:
            return None
        
        node = instance["node"]
        comfyui_url = f"https://ui-{node}"
        
        # Validate URL format - check for common typos in the node name
        if "comput3.ai" not in node:
            comfyui_url = f"https://ui-{node}.comput3.ai"
            
        logger.info(f"🌐 ComfyUI URL: {comfyui_url}")
        
        # Test connection to URL
        try:
            test_response = requests.head(comfyui_url, timeout=5)
            if test_response.status_code >= 400:
                logger.warning(f"⚠️ ComfyUI URL test failed with status code: {test_response.status_code}")
        except requests.exceptions.RequestException as e:
            logger.warning(f"⚠️ ComfyUI URL may be incorrect, connection test failed: {str(e)}")
            logger.warning("⚠️ Will attempt to use the URL anyway, but there may be connection issues")
                
        return comfyui_url 